### PR TITLE
[DN] Creates a StoryRequest and use cases for loading,searching and g…

### PR DESCRIPTION
**Pull request by florina-muntenescu**
…etting stories

### CODE REVIEW

File: core/src/test/java/io/plaidapp/core/designernews/data/stories/StoriesRepositoryTest.kt [hunk 1]
```diff
@@ -180,23 +113,23 @@ class StoriesRepositoryTest {
...
-        // Then an Error is reported
+        // Then error is returned
         assertNotNull(result)
         assertTrue(result is Result.Error)
     }
 
     @Test
     fun getStory_whenSearchSucceeded() = runBlocking {
-        // Given that a search has been performed successfully and data cached
+        // Given that a invoke has been performed successfully and data cached
```
Comment (ataulm) [inline core/src/test/java/io/plaidapp/core/designernews/data/stories/StoriesRepositoryTest.kt]: :eyes:

File: core/src/test/java/io/plaidapp/core/designernews/data/stories/StoriesRepositoryTest.kt [hunk 2]
```diff
@@ -39,135 +34,73 @@ import java.util.GregorianCalendar
...
-        // The correct callback was called
-        assertTrue(sourceLoaded)
+        // The correct data is returned
+        assertEquals(Result.Success(stories), data)
     }
 
     @Test
     fun search_withError() = runBlocking {
-        // Given that an error is returned for a specific query and page search
+        // Given that an error is returned for a specific query and page invoke
```
Comment (ataulm) [inline core/src/test/java/io/plaidapp/core/designernews/data/stories/StoriesRepositoryTest.kt]: :eyes: do a quick search for "invoke" on this PR

File: designernews/src/main/java/io/plaidapp/designernews/domain/GetStoryUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import io.plaidapp.core.designernews.data.stories.model.Story
+import io.plaidapp.core.designernews.data.stories.model.StoryResponse
+import io.plaidapp.core.designernews.data.stories.model.toStory
+
+/**
+ * Use case that gets a [StoryResponse]s from [StoriesRepository] and transforms it to [Story]
+ */
+class GetStoryUseCase(private val storiesRepository: StoriesRepository) {
+
+    operator fun invoke(id: Long): Result<Story> {
```
Comment (ataulm) [inline designernews/src/main/java/io/plaidapp/designernews/domain/GetStoryUseCase.kt]: noice http://joshskeen.com/kotlins-invoke-operator/

so if you had `Provider#provideGetStoryUseCase() : GetStoryUseCase`, you could:

```kotlin
val id = 1
val getStory = provider.provideGetStoryUseCase()
val result = getStory(id)
```
It looks like a nice pattern for interactors/use cases that are named with verbs/commands :raised_hands:

File: designernews/src/main/java/io/plaidapp/designernews/ui/story/StoryViewModel.kt [hunk 4]
```diff
@@ -42,7 +42,7 @@ class StoryViewModel(
     val story: Story
 
     init {
-        val result = storiesRepository.getStory(storyId)
+        val result = getStoryUseCase(storyId)
```
Comment (ataulm) [inline designernews/src/main/java/io/plaidapp/designernews/ui/story/StoryViewModel.kt]: :raised_hands: one day I'll scroll a little ahead before commenting/asking.

File: core/src/main/java/io/plaidapp/core/designernews/data/stories/StoriesRepository.kt [hunk 5]
```diff
@@ -16,50 +16,31 @@
...
-        page: Int,
-        callback: LoadSourceCallback
-    ) {
-        val jobId = "$query::$page"
-        parentJobs[jobId] = launchRequest(query, page, callback, jobId) {
-            remoteDataSource.search(query, page)
+    private suspend fun getData(
+        request: suspend () -> Result<List<StoryResponse>>
+    ): Result<List<StoryResponse>> {
+        val result = request.invoke()
```
Comment (nickbutcher) [inline core/src/main/java/io/plaidapp/core/designernews/data/stories/StoriesRepository.kt]: Just `request()`

File: core/src/test/java/io/plaidapp/core/designernews/domain/SearchStoriesUseCaseTest.kt [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            override fun sourceLoaded(result: List<PlaidItem>?, page: Int, source: String) {
+                // Then the correct data is received
+                sourceLoaded = true
+                assertEquals(stories, result)
+                assertEquals(1, page)
+                assertEquals(query, source)
+            }
+
+            override fun loadFailed(source: String) {
+                Assert.fail("Search shouldn't have failed")
```
Comment (nickbutcher) [inline core/src/test/java/io/plaidapp/core/designernews/domain/SearchStoriesUseCaseTest.kt]: import method as elsewhere.

File: designernews/src/main/java/io/plaidapp/designernews/ui/story/StoryViewModel.kt [hunk 7]
```diff
@@ -33,7 +33,7 @@ import kotlinx.coroutines.experimental.withContext
  */
 class StoryViewModel(
     storyId: Long,
-    storiesRepository: StoriesRepository,
+    getStoryUseCase: GetStoryUseCase,
```
Comment (nickbutcher) [inline designernews/src/main/java/io/plaidapp/designernews/ui/story/StoryViewModel.kt]: [Can be in a follow up] lets drop the `*UseCase` suffix on instances e.g. just `getStory` here

**Pull request merged**